Podcast/Player.swift

```diff
@@ -9,59 +9,189 @@
+                self?.delegate?.playerDidUpdateTime()
+            })
+        }
     }
     
+    /// Pauses playback of the current track and sets the Player status to .paused if successful
     func pause() {
-        player?.pause()
+        if let player = currentAVPlayer {
+            player.pause()
+            playerStatus = .paused
+        }
     }
     
+
+    /// Toggles playback of the Player. If the player is in the preparingToPlay state and cannot yet play,
+    /// this toggles whether the player will autoplay once it is ready.
     func togglePlaying() {
-        isPlaying ? pause() : play()
+        switch playerStatus {
+        case .empty:
+            break
+        case .preparingToPlay:
+            shouldAutoPlay = !shouldAutoPlay
+        case .playing:
+            pause()
+        case .paused:
+            play()
+        case .failed:
+            break
+        }
     }
     
-    // MARK: - AVAudioPlayerDelegate
+    /// Skips a variable amount of seconds in the current audio track. Skips forward or backward based on a positive or negative seconds value.
+    /// - Parameters:
+    ///   - seconds: the amount of time in seconds to skip. If negative, skips backward.
+    func skip(seconds: Float64) {
+        if let player = currentAVPlayer {
+            let newTime = CMTimeAdd(player.currentTime(), CMTimeMakeWithSeconds(seconds, Int32(NSEC_PER_SEC)))
+            player.seek(to: newTime)
+            self.delegate?.playerDidUpdateTime()
+        }
+    }
     
     
-    func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
-        pause()
+    /// Returns the progress for the Player's current track.
+    /// - Returns: a Float in the range [0.0,1.0] specifying what the current location of playback is relative to the entire track.
+    func getProgress() -> Float {
+        if playerStatus == .empty || playerStatus == .failed || currentAVPlayer == nil || currentAVPlayer?.currentItem == nil {
+            return 0.0
+        }
+        return Float((currentAVPlayer?.currentItem?.currentTime().seconds)!) / Float((currentAVPlayer?.currentItem?.duration.seconds)!)
+    }
+    
+    /// Sets the progress of the Player's current track
+    /// - Parameters:
+    ///   - progress: the value to set where the Player should begin playback. Should be in range [0.0,1.0], but will be set to 0.0 if less than 0.0 and 1.0 if greater than 1.0
+    func setProgress(progress: Float) {
+        if playerStatus == .empty || playerStatus == .failed {
+            return
+        }
+        
+        if progress < 0.0 {
+            setProgress(progress: 0.0)
+        } else if progress > 1.0 {
+            setProgress(progress: 1.0)
+        }
+        
+        let duration = currentAVPlayer?.currentItem?.duration
+        let newTime = progress * Float((duration?.seconds)!)
```

Comment:
duration may be nil if currentAVPlayer is nil so to safeguard against that, do `if let duration = currentAVPlayer?.currentItem?.duration`